aa-user/protocol.go

```diff
@@ -28,20 +28,20 @@ const (
 
 // Protocol stores all the RMC method handlers for the AAUser protocol and listens for requests
 type Protocol struct {
-	Server                       *nex.Server
-	registerApplicationHandler   func(err error, packet nex.PacketInterface, callID uint32, titleID uint64) uint32
-	unregisterApplicationHandler func(err error, packet nex.PacketInterface, callID uint32, titleID uint64) uint32
-	setApplicationInfoHandler    func(err error, packet nex.PacketInterface, callID uint32, applicationInfo []*aauser_types.ApplicationInfo) uint32
-	getApplicationInfoHandler    func(err error, packet nex.PacketInterface, callID uint32) uint32
+	Server                nex.ServerInterface
+	RegisterApplication   func(err error, packet nex.PacketInterface, callID uint32, titleID uint64) (*nex.RMCMessage, uint32)
+	UnregisterApplication func(err error, packet nex.PacketInterface, callID uint32, titleID uint64) (*nex.RMCMessage, uint32)
+	SetApplicationInfo    func(err error, packet nex.PacketInterface, callID uint32, applicationInfo []*aauser_types.ApplicationInfo) (*nex.RMCMessage, uint32)
+	GetApplicationInfo    func(err error, packet nex.PacketInterface, callID uint32) (*nex.RMCMessage, uint32)
 }
 
 // Setup initializes the protocol
 func (protocol *Protocol) Setup() {
-	protocol.Server.On("Data", func(packet nex.PacketInterface) {
-		request := packet.RMCRequest()
+	protocol.Server.OnData(func(packet nex.PacketInterface) {
+		request := packet.RMCMessage()
 
-		if request.ProtocolID() == ProtocolID {
-			switch request.MethodID() {
+		if request.ProtocolID == ProtocolID {
+			switch request.MethodID {
 			case MethodRegisterApplication:
 				go protocol.handleRegisterApplication(packet)
```

Comment:
Commenting here, but applies everywhere

Do we need to make new threads here? Or can we make it run all on the same thread?

Reply:
These are already running inside a goroutine so yeah, there's no reason for a new one to be made in this case (this is what I was getting at the other day)

Forgot to remove these, my bad